icubam/backoffice/handlers/users.py

```diff
@@ -39,50 +40,111 @@ def get(self):
+    icus = self.store.get_icus()
+    output = []
+    for icu in icus:
+      dto = icu.to_dict()
+      dto["selected"] = icu.icu_id in subset_icus_ids
+      output.append(dto)
+    return sorted(output, key = lambda x: x['name'])
+
+  """Prepares the DTO comparing ALL icus against the ones from the user."""
+  def prepare_icus_for_user(self, user):
+    return self.prepare_icus([user_icu.icu_id for user_icu in user.icus])
+
   @tornado.web.authenticated
   def get(self):
     userid = self.get_query_argument('id', None)
     user = None
     if userid is not None:
-      user = self.db.get_user(userid)
-    # TODO(fred): load the ICUs to show in the form.
+      user = self.store.get_user(userid)
+
     user = user if user is not None else store.User()
-    self.render("user.html", icus=[], user=user, error="")
+    icus_dto = self.prepare_icus_for_user(user)
+    self.render(
+      "user.html", icus=icus_dto, user=user, error=False,
+      allow_admin=self.user.is_admin)
 
-  def error(self, error_message, user):
-    self.render("user.html", icus=[], user=user, error=error_message)
+  def error(self, user, icus):
+    self.render(
+      "user.html", icus=icus, user=user, error=True,
+      allow_admin=self.user.is_admin)
+
+  """Returns the ICUs we need to add to/remove from the user. """
+  def get_icus_to_store(self, user):
+    form_icus = self.get_body_arguments("icus", [])
+    form_icus = set([int(icu) for icu in form_icus])
+    user_icus = set([user_icu.icu_id for user_icu in user.icus])
+    add =  list(form_icus.difference(user_icus))
+    remove = list(user_icus.difference(form_icus))
+    return add, remove
+
+  """Prepares the DTO comparing ALL icus against the ones from the form."""
+  def prepare_icus_for_error(self):
+    form_icus = self.get_body_arguments("icus", [])
+    return self.prepare_icus([int(icu) for icu in form_icus])
 
   @tornado.web.authenticated
   def post(self):
-    # We first create a potential user object, because if the validation fails,
-    # we send it back to the client and the client can re-fill the form with
-    # the info that was sent.
-    user = User(name=self.get_body_argument("name", ""),
-      telephone=self.get_body_argument("telephone", ""),
-      email=self.get_body_argument("email", ""),
-      password_hash=self.store.get_password_hash(
-        self.get_body_argument("password", "")))
-
-    if not user.name:
-      return self.error("Missing or invalid user name", user)
-
-    if not user.email:
-      return self.error("Missing or invalid user email", user)
+    user_dict = {}
+    for key in ["name", "email", "telephone"]:
+      user_dict[key] = self.get_body_argument(key, "")
+
+    # Password is not required when editing the user.
+    password = self.get_body_argument("password", "")
+    if password:
+      user_dict["password_hash"] = self.store.get_password_hash(password)
+
+    uid = self.get_body_argument("user_id", None)
+    if uid:
```

Comment:
better use the `is not None` syntax IMO, since other things could return False (0, "", etc.)

Reply:
yeah, I thought about it after going to bed yesterday.

Reply:
there's a small little thing that I don't understand well with hidden fields (they come empty always, not None). I will check that later, so I had to keep the same check here.

Reply:
ok!

Reply:
probably because it is body argument. so there is no such thing as None. it s all strings